ext/webserial/01_webserial.js

```diff
@@ -0,0 +1,397 @@
+      webidl.illegalConstructor();
+    }
+
+    // deno-lint-ignore require-await
+    async getPorts() {
+      webidl.assertBranded(this, Serial);
+      const ports = core.opSync("op_webserial_get_ports");
+
+      // TODO(@crowlKats): maybe cache ports?
+      return ports.map((port) => {
+        const serial = webidl.createBranded(SerialPort);
+        serial[_name] = port.name;
+        serial[_info] = {
+          usbVendorId: info?.[0],
+          usbProductId: info?.[1],
+        };
+        return serial;
+      });
+    }
+
+    /*async requestPort(options = {}) {
+      webidl.assertBranded(this, Serial);
+      const prefix = "Failed to execute 'requestPort' on 'Serial'";
+      webidl.requiredArguments(arguments.length, 1, { prefix });
+      options = webidl.converters.SerialPortRequestOptions(options, {
+        prefix,
+        context: "Argument 1",
+      });
+
+      if (options.filters) {
+        for (const filter of options.filters) {
+          if (!("usbVendorId" in filter)) {
+            throw TypeError();
+          }
+        }
+      }
+    }*/
+  }
+
+  const _state = Symbol("[[state]]");
+  const _bufferSize = Symbol("[[bufferSize]]");
+  const _readable = Symbol("[[readable]]");
+  const _readFatal = Symbol("[[readFatal]]");
+  const _writable = Symbol("[[writable]]");
+  const _writeFatal = Symbol("[[writeFatal]]");
+  const _pendingClosePromise = Symbol("[[pendingClosePromise]]");
+  const _rid = Symbol("[[rid]]");
+  const _name = Symbol("[[name]]");
+  const _info = Symbol("[[info]]");
+
+  class SerialPort {
+    [_state] = "closed";
+    [_bufferSize] = undefined;
+    [_readable] = null;
+    [_readFatal] = false;
+    [_writable] = null;
+    [_writeFatal] = false;
+    [_pendingClosePromise] = null;
+
+    [_rid];
+    [_name];
+    [_info];
+
+    get readable() {
+      webidl.assertBranded(this, SerialPort);
+      if (this[_readable] !== null) {
+        return this[_readable];
+      }
+      if (this[_state] !== "opened") {
+        return null;
+      }
+      if (this[_readFatal]) {
+        return null;
+      }
+
+      const stream = new ReadableStream({
+        pull: async (controller) => {
+          const buf = new Uint8Array(controller.desiredSize);
+          await core.opAsync("op_webserial_read", this[_rid], buf); // TODO(@crowlKats): errors
+          controller.enqueue(buf);
+        },
+        cancel: () => {
+          // TODO(@crowlKats): Invoke the operating system to discard the contents of all software and hardware receive buffers for the port.
+          this.#handleClosingReadable();
+        },
+      }, {
+        highWaterMark: this[_bufferSize],
+        size: (chunk) => chunk.byteLength,
+      });
+
+      this[_readable] = stream;
+      return stream;
+    }
+
+    #handleClosingReadable() {
+      this[_readable] = null;
+      if (this[_writable] === null && this[_pendingClosePromise] !== null) {
+        this[_pendingClosePromise].resolve(undefined);
```

Comment:
Do these resolve calls actually work somehow? Promise does not have a method to resolve it from the outside, and these Promises seem to be created with the normal primordials Promise ie. globalThis Promise.

Reply:
Probably not. as stated, PR was extremely WIP when opened, and now is quite stale.